src/lib/unlang/xlat.c

```diff
@@ -304,8 +308,95 @@ static unlang_action_t unlang_xlat_repeat(rlm_rcode_t *p_result, request_t *requ
 	xlat_action_t			xa;
 	xlat_exp_head_t const		*child = NULL;
 
+	/*
+	 *	If the xlat is a function with a method_env, expand it before calling the function.
+	 */
+	if ((state->exp->type == XLAT_FUNC) && state->exp->call.func->method_env) {
```

Comment:
Move this to a separate static inline.  It's all nicely self contained and does one thing, so it's clearer to put it in a separate function.